### ISSUE TITLE
[RFE] add button for watching an issue

### ISSUE THREAD

**Issue by mathstuf** [slapperwan/gh4a#215]
This way I can watch an issue without making a dummy comment (see #129).

**Comment by bhack** [comment]
Is not this https://developer.github.com/v3/activity/notifications/#set-a-thread-subscription?

**Comment by myrdd** [comment]
@bhack Looks like the API interface we are looking for.

**Comment by maniac103** [comment]
I think I found that one earlier; what I missed was a way to find out the thread ID.

**Comment by alorma** [comment]
There's no way to get the thread ID, until you get the notification...

Asked to github team.

**Comment by maniac103** [comment]
Thanks. So it looks like watching an issue indeed isn't possible with the current API.

**Comment by mammuth** [comment]
So, were there any related API changes in the last year so subscribing might work?

**Comment by mathstuf** [comment]
I don't see any relevant API endpoints still :( .

**Comment by sbrl** [comment]
Any status updates on this? (I'm posting this comment primarily to subscribe to this issue lol)

**Comment by mammuth** [comment]
I just talked to the Github support some days ago, just about this issue. There is still no way to retrieve the thread ID without a notification. They know about this but they were not able to provide me a roadmap or information on whether / when this might be resolved.

**Comment by sbrl** [comment]
That's annoying. I could really use it, too 😕

**Comment by Bluscream** [comment]
How does the browser get the thread id?

**Comment by maniac103** [comment]
Github's web frontend doesn't use the official API, but a combination of server-side rendered pages and an unofficial API. They simply have more possibilities.

**Comment by Bluscream** [comment]
Can't octodroid e.simply emulate that?

**Comment by maniac103** [comment]
No, it can't. We'll stick to the API, everything else us not supportable.

**Comment by jancborchardt** [comment]
In the notifications view though there is a menu entry to "Unsubscribe". Couldn't the same "Subscribe"/" Unsubscribe" toggle be shown in the 3-dot-menu of an issue view?

**Comment by Tunous** [comment]
Nope, notifications API does provide subscription id but it's not available when accessing the issue by itself and there is no way to request this separately either.

**Comment by laughedelic** [comment]
> notifications API does provide subscription id but it's not available when accessing the issue by itself

It is available (at least now) in the response for the [`GET /repos/:owner/:repo/issues/:number`](https://developer.github.com/v3/issues/#get-a-single-issue) request (see the `id` parameter). See for example [this issue](https://api.github.com/repos/slapperwan/gh4a/issues/215): 
```
  "id": 43842983,
  "number": 215,
```

As far as I understand this `id` is the "thread id" that can be used to subscribe/unsubscribe through the notifications API.

**Comment by maniac103** [comment]
> As far as I understand this id is the "thread id" that can be used to subscribe/unsubscribe through the notifications API.

How did you come to this conclusion? I don't think it's the case. As an example:

- curl https://api.github.com/repos/slapperwan/gh4a/issues/777
```
[...]
  "id": 265823140,
  "number": 777,
[...]    
```

- curl -H "Authorization: Token XXX" https://api.github.com/notifications
```
[...]
    "id": "267086772",
    "unread": true,
    "reason": "mention",
    "updated_at": "2017-10-20T13:58:47Z",
    "last_read_at": "2017-10-16T18:01:45Z",
    "subject": {
      "title": "Line comments being typed lost on portrait/landscape rotation",
      "url": "https://api.github.com/repos/slapperwan/gh4a/issues/777",
      "latest_comment_url": "https://api.github.com/repos/slapperwan/gh4a/issues/comments/338214630",
      "type": "Issue"
    },
[...]
    "subscription_url": "https://api.github.com/notifications/threads/267086772/subscription"
[...]
```

So - AFAICT - issue ID and subscription ID is *not* the same thing. Consequently, 
```
curl -X PUT -d subscribed=false -d ignored=false -H "Authorization: Token XXX" https://api.github.com/notifications/threads/265823140/subscription
```

yields
```
{
  "message": "Must have admin rights to Repository.",
  "documentation_url": "https://developer.github.com/v3/activity/notifications/#set-a-thread-subscription"
}
```

so it seems to have interpreted the ID differently.

**Comment by laughedelic** [comment]
You are right! I shouldn't have interpreted it this way before testing as you did 👍

**Comment by evaryont** [comment]
It seems like you could query `https://github.com/USERNAME/REPONAME/issues/ISSUE_ID/show_partial?partial=issues%2Ftitle` and parse out data-channel (looks like "issue:THREAD_ID") from there. Definitely unofficial API, however.

**Comment by gsantner** [comment]
Has somebody started on integrating the suggested workaround?

**Comment by mathstuf** [comment]
It looks like the GraphQL API [has support](https://developer.github.com/v4/mutation/updatesubscription/) for changing subscriptions.

**Comment by KaKi87** [comment]
Hello, any news on this ?

**Comment by mathstuf** [comment]
#605 seems to be the relevant issue there. The app would need to be ported to use GraphQL. https://blog.apollographql.com/launching-apollo-graphql-on-android-40ee0b5789bd?gi=46c44c646e07 seems like a promising way to do it (though I just did a simple search and other libraries may do it better in the time since that post).

**Comment by KaKi87** [comment]
Or the app could query the same endpoint the client does, regardless of it's official or not.

**Comment by mathstuf** [comment]
GitHub is not shy about breaking undocumented APIs. The thread ID scraping could break, the endpoint parameter could change, etc. They may also block apps using them (especially now that they have a supported endpoint for it). It's better to use the correct APIs for these kinds of things.

**Comment by KaKi87** [comment]
So you prefer to wait for the more complicated implementation and have nothing in the meantime instead of implementing something when we can ?

**Comment by maniac103** [comment]
> when we can 

*We* can? Did I miss a pull request? :-P

**Comment by KaKi87** [comment]
Sorry I code JS not Java 😅
What I mean : you said that GraphQL migration would be the most complicated option, so using the user endpoint would be easier

**Comment by mathstuf** [comment]
> So you prefer to wait for the more complicated implementation and have nothing in the meantime instead of implementing something when we can ?

As said above, if it were so easy, why isn't it already done by now? The main issue here is gathering the data for the query since the data is spread around the *HTML* page (not any API object you get back for an issue).

**Comment by iamtalhaasghar** [comment]
Any updates on this issue?

**Comment by mathstuf** [comment]
AFAICT, there's still no API for this. :(

**Comment by mathstuf** [comment]
Looks like GraphQL might have something now: https://docs.github.com/en/graphql/reference/interfaces#subscribable

**Comment by Cwpute** [comment]
I'm not sure if what's available from GraphQL would allow that, but if this gets implemented, i'd appreciate it if we could personalize the subscription as well (only on closing/reopening the issue, only on mentions etc…).
This is especialy useful for those who keep an eye on projects that are missing key features or have workflow-breaking bugs, in hope they get adressed, so that they can introduce it to their own stack.

**Comment by mathstuf** [comment]
> i'd appreciate it if we could personalize the subscription as well

This is a Github feature that (AFAIK) is completely missing. All gh4a can do is change the subscription setting (ignore/unsub/sub) as that is all that Github exposes. gh4a might be able to filter notifications it forwards to Android, but that's about it.

**Comment by xa4hf8** [comment]
@Mynacol 
> I wanted a convenient watch/unsubscribe button in 
this app and found this old issue. Can someone check if there's an API method nowadays?

https://github.com/copilot/share/4a1e109a-0b24-80f7-a951-da0ac47f2817
